### DISCUSSION TITLE
Why doesn't C# realise that this reference type variable is guaranteed to be set?

### DISCUSSION THREAD

**Discussion by jez9999** [dotnet/csharplang#7675] [General] [answered]
I have the following code:

```
private TestClass? getTestClass() {
    return new TestClass();
}

private void testMethod() {
    var testClass = getTestClass();
    var testBool = true;
    if (testBool) {
        testClass = new TestClass();
    }
    var someInt = testClass.IntMember;
}
```

The last line gives me a warning with nullable reference types: Dereference of a possibly null reference.

Why is this a possibly null reference? `testBool` must be true, the if block must get run, and the variable must be initialized to non-null. Oddly, if I change to `if (true)` it works. Why doesn't `if (always-true-bool-variable)`?

**Comment by HaloFour** [comment, marked as answer]
Because nullable flow analysis does not track boolean conditions across variable assignments.  I believe that there is a proposal to expand that to support trivial cases like this one, but there will always be a limit to how far flow analysis will be willing to interpret and evaluate the code to determine which code paths may or may not execute.

**Reply by jez9999** [reply]
Yeah, it seems pretty darn limited.  I don't mind them encouraging people to use nullable reference types but not having stuff like this means putting exclamation marks all over the place to tell the compiler that things are OK, which is not good.

**Reply by HaloFour** [reply]
Or don't write code like your example above.  There's no point to have a Boolean variable assigned to a constant that is then used in a condition.  If the condition can only ever be true or false, get rid of it.  Any realistic code won't be that trivial.  The team isn't going to sink tons and tons of effort into handling esoteric and pointless edge cases.

**Reply by jez9999** [reply]
Yeah, in practice my code obviously is a bit more complex.  There are several `if` statements that throw if the variable is null to the effect that code flow analysis should be able to determine that the variable will definitely be non-null.  I just simplified it down into that.  (why do people bother simplifying things down into basic examples?  People just come along and say "don't do that basic example" as if it's not a darned example)

**Reply by HaloFour** [reply]
Given the answer has to do with the complexity of the method, that seems completely valid.  I can only work with the information as provided.  Decrease the indirection and flow analysis should work fine.

**Reply by jez9999** [reply]
I can't decrease the indirection.  I have various `if` blocks, one of which is guaranteed to set the variable based on a certain condition.  What am I meant to do?

**Reply by HaloFour** [reply]
I can't give concrete advice without concrete examples.  I'd suggest that the majority of situations are covered by existing flow analysis, or can be through modest refactoring.  If your example can't then that is a case where you can inform the compiler via the null-forgiveness operator or the like.

**Reply by CyrusNajmabadi** [reply]
> which is not good.

Nullable analysis is best effort. And there is an unending amount of flow analysis that could be done to make more precise understanding of the code. No matter where you draw the line, some cases will fall through.  That's literally why there is a suppression operator.

**Reply by tannergooding** [reply]
Notably there is also `Debug.Assert(testClass is not null);`, which I personally prefer over suppression

It does the same job, but is more readable/visible, provides a convenient place to breakpoint if it fails, and a convenient place to add any annotation/comment as to why it can’t be null for future readers (including yourself 3mo later)

**Reply by jez9999** [reply]
@tannergooding Unfortunately that doesn't work with the mechanism I'm using, which is to initialize the variable as a side-effect of a lambda passed to an awaited method that executes it.  It won't let me do the `Debug.Assert` as it's sure that `testClass` is an uninitialized variable.

**Reply by jez9999** [reply]
@tannergooding It also doesn't work in this example if, instead of `var testClass = getTestClass();`, you simply write `TestClass testClass;`.
